crates/wasmtime/src/runtime/func/typed.rs

```diff
@@ -17,17 +21,19 @@ use wasmtime_runtime::{
 ///
 /// This structure is created via [`Func::typed`] or [`TypedFunc::new_unchecked`].
 /// For more documentation about this see those methods.
-#[repr(transparent)] // here for the C API
 pub struct TypedFunc<Params, Results> {
     _a: marker::PhantomData<fn(Params) -> Results>,
+    ty: FuncType,
```

Comment:
We could remove the dynamic downcasting checks from the typed function API and instead explore (in a follow up PR) implementing `WasmTy` for `TypedFunc<P, R>` such that you could do

```rust
// succeeds if typeof(f) <: (func (param i64) (result (ref $t)))
// where $t = (type (func (param funcref) (result (ref null $u))))
// and $u = (type (func (result i32)))
let f = instance.get_typed_func::<
    i64,
    TypedFunc<
        Option<Func>,
        Option<TypedFunc<(), i32>>,
    >,
>(&store)?;
```

Chris was suggesting this to me yesterday, and I think it should be doable. It would allow us to completely remove dynamic checks for calling concrete function references, the way `Func::typed` works today.

What do you think of that?

Reply:
Oh now that's clever, I really like that!

Would that remove the need for this field though? I thought that this was used additionally as part of `call_raw` where the type is needed to know what the params/results are?